subsys/net/lib/sockets/sockets_service.c

```diff
@@ -0,0 +1,301 @@
+
+			break;
+		}
+	}
+
+	k_mutex_unlock(&lock);
+
+	return ret;
+}
+
+static struct net_socket_service_desc *find_svc_and_event(
+	struct zsock_pollfd *pev,
+	struct net_socket_service_event **event)
+{
+	STRUCT_SECTION_FOREACH(net_socket_service_desc, svc) {
+		for (int i = 0; i < svc->pev_len; i++) {
+			if (svc->pev[i].event.fd == pev->fd) {
+				*event = &svc->pev[i];
+				return svc;
+			}
+		}
+	}
+
+	return NULL;
+}
+
+static void service_callback_sync(struct k_work *work)
+{
+	struct net_socket_service_event *pev =
+		CONTAINER_OF(work, struct net_socket_service_event, work);
+	struct net_socket_service_event ev = *pev;
+
+	ev.callback(&ev.work);
+}
+
+static int call_work(struct zsock_pollfd *pev, struct k_work_q *work_q,
+		     struct k_work *work)
+{
+	int ret = 0;
+
+	if (work->handler == NULL) {
+		/* Synchronous call */
+		service_callback_sync(work);
+	} else {
+		/* Mark the global fd non pollable so that we do not
+		 * call the callback second time.
+		 */
+		pev->fd = -1;
+
+		if (work_q != NULL) {
+			ret = k_work_submit_to_queue(work_q, work);
+		} else {
+			ret = k_work_submit(work);
+		}
+
+		k_yield();
+	}
+
+	return ret;
+
+}
+
+static int trigger_work(struct zsock_pollfd *pev)
+{
+	struct net_socket_service_event *event;
+	struct net_socket_service_desc *svc;
+
+	svc = find_svc_and_event(pev, &event);
+	if (svc == NULL) {
+		return -ENOENT;
+	}
+
+	event->svc = svc;
+
+	return call_work(pev, svc->work_q, &event->work);
+}
+
+/* We do not set the user callback to our work struct because we need to
+ * hook into the flow and set the related fd to -1 so that the next poll
+ * round will not notice it and call the callback again while we are
+ * servicing the callback.
+ */
+void net_socket_service_callback(struct k_work *work)
+{
+	struct net_socket_service_event *pev =
+		CONTAINER_OF(work, struct net_socket_service_event, work);
+	struct net_socket_service_desc *svc = pev->svc;
+	struct net_socket_service_event ev = *pev;
+
+	ev.callback(&ev.work);
+
+	/* Copy back the socket fd to the global array (because we marked
+	 * it as -1 when triggering async work.
+	 */
+	for (int i = get_idx(svc); i < get_idx(svc) + svc->pev_len; i++) {
+		for (int j = 0; j < svc->pev_len; j++) {
+			ctx.events[i] = svc->pev[j].event;
+		}
+	}
+}
+
+static void socket_service_thread(void)
+{
+	int ret, i, fd, count = 0;
+	eventfd_t value;
+
+	STRUCT_SECTION_COUNT(net_socket_service_desc, &ret);
+	if (ret == 0) {
+		NET_INFO("No socket services found, service disabled.");
+		goto fail;
+	}
+
+	/* Create contiguous poll event array to enable socket polling */
+	STRUCT_SECTION_FOREACH(net_socket_service_desc, svc) {
+		get_idx(svc) = count + 1;
+		count += svc->pev_len;
+	}
+
+	if (count > (ARRAY_SIZE(ctx.events) - 1)) {
+		NET_WARN("You have %d services to monitor but "
+			 "%d poll entries configured.",
+			 count, ARRAY_SIZE(ctx.events) - 1);
+		NET_WARN("Consider increasing value of %s to %d",
+			 "CONFIG_NET_SOCKETS_POLL_MAX", count);
```

Comment:
Note to self, the eventfd was missing so the poll max needs to be at least `count + 1`